### ISSUE TITLE
Delete mutations seem to get stuck semi-randomly

### ISSUE THREAD

**Issue by veqryn** [ClickHouse/ClickHouse#41546]
**Describe what's wrong**
The entire time our company has used Clickhouse (at least 4 years now, maybe a lot more in one of our acquisitions), it has always had problems with deleting data.  Most of the time, the problem would surface as delete mutation processes that are running but appear to be stuck, even after many days.  Sometimes multiple stuck deletes would start accumulating and would somehow put the table or even the whole cluster into a bad state.  All of our engineers are either paranoid about running deletes, or are new and told they should be paranoid about it.  We even specifically engineer our products to avoid deleting data if at all possible.

I have had 5 out of my last 10 delete queries get stuck or frozen this week.
The delete queries are valid and simple, and killing the process followed by re-running them results in their successful completion.

**Does it reproduce on recent release?**
We are running version `21.8.12.29` of clickhouse.

**How to reproduce**
Our clickhouse cluster has 24 beefy nodes.

Example table:
```sql
CREATE TABLE analytics.xyz_local ON CLUSTER analytics_cluster (
    created          DateTime,
    eventday         Date,
    user_id          UInt32,
    ttldays          UInt16                 DEFAULT 548
    -- About 40 other columns...
)
    ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/analytics.xyz_local', '{replica}')
    PARTITION BY toYYYYMM(eventday)
    ORDER BY (user_id, eventday, date_trunc('hour', created), date_trunc('minute', created), intHash64(created))
    TTL eventday + toIntervalDay(ttldays) DELETE
    SAMPLE BY intHash64(created);

CREATE TABLE IF NOT EXISTS analytics.xyz ON CLUSTER analytics_cluster AS analytics.xyz_local ENGINE = Distributed(analytics_cluster, analytics, xyz_local, rand());
```

We have around 200 billion rows of data in this table, spanning the last 1.5 years, taking up about 650 Gigs of space.

Examples of the kinds of delete queries that have become stuck:
```sql
alter table analytics.xyz_local on cluster analytics_cluster DELETE WHERE eventday = '2022-05-01';

alter table analytics.xyz_local on cluster analytics_cluster DELETE WHERE eventday >= '2022-05-16' and eventday < '2022-06-01';

alter table analytics.xyz_local on cluster analytics_cluster DELETE WHERE eventday = '2022-06-26' and created >= '2022-06-26 06:00:00' and created < '2022-06-26 12:00:00' and `source` = 'foobar' and activity = 'forward';

alter table analytics.xyz_local on cluster analytics_cluster DELETE WHERE eventday < '2022-06-01' and source <> 'ecm';

alter table analytics.xyz_local on cluster analytics_cluster DELETE WHERE eventday < '2022-08-01';
```

We've also had very similar deletion queries run to success on the first try.  It seems pretty random which ones succeed and which ones won't.
From what I've noticed, it seems like the more data there is to delete, the more likely the query will become stuck.

All of those queries, except for the last one, have been killed and re-run successfully.

I've let the last one keep running, for 4 days now, so that maybe we can debug it.

It has been stuck with 401 parts left to do, for the majority of that time:
```sql
select database, table, command, sum(parts_to_do) from clusterAllReplicas(analytics_cluster, system, mutations) where is_done <> 1 group by database, table, command order by database, table, command;
```
`analytics	xyz_local	DELETE WHERE eventday < '2022-08-01'	401`

It is stuck on 13 out of our 24 nodes:
```sql
select fqdn(), * from clusterAllReplicas(analytics_cluster, system, mutations) where is_done <> 1 order by database, table, command;
```

```
clickhouse-us-east-1b-11.something	analytics	xyz_local	0000000541	DELETE WHERE eventday < '2022-08-01'	2022-09-17 22:58:18.000	['197001','202104','202105','202106','202107','202108','202109','202110','202111','202112','202201','202202','202203','202204','202205','202206','202207','202208','202209']	[6022,6968,7123,7155,57387,57294,56817,54368,60081,75373,80569,77359,102399,106376,380669,990256,1061659,1442959,834792]	['202104_3043_5805_5_6967','202105_2829_5820_5_7122','202106_0_2841_5_7154','202106_2842_5724_5_7154','202107_0_56859_18_57386','202107_56860_56860_0_57386','202107_56861_56861_0_57386','202108_0_51581_95_57293','202108_51582_54075_6_57293','202108_54076_55560_6_57293','202108_55561_56353_6_57293','202108_56354_56761_8_57293','202108_56762_56768_2_57293','202109_55818_56011_7_56816','202109_56012_56287_17_56816','202109_56288_56291_1_56816','202110_0_33280_11_54367','202110_33281_43054_8_54367','202110_43055_51784_8_54367','202110_51785_53072_8_54367','202110_53073_54328_33_54367','202112_0_74991_40_75372','202201_0_13325_53_80568','202201_54810_61520_43_80568','202203_0_45755_176_102398','202206_742652_922030_308_990255','202207_1043047_1045234_5_1061658','202207_1045235_1045797_4_1061658','202207_1046278_1061026_285_1061658']	29	0		1970-01-01 00:00:00.000	
-- and 12 more similar lines to above ^
```

There don't appear to be any related errors in the system.errors table.
There were probably around 4000 parts to be done before the query was ran.  Just a guess though.
Our DBA checked but couldn't find any reasons why the delete mutation was stuck.  This is on our production servers, and noone else has reported any other problems or anomalies.

**Expected behavior**
Delete mutations should run to completion.

**Comment by den-crane** [comment]
>'202104_3043_5805_5_6967','202105_2829_5820_5_7122','202106_0_2841_5_7154'

Check that this parts really exists in CH (system.parts). Most probably they are orphans which exists in zookeper but not in CH and they should be removed from ZK. 
If this partitions 202104, 202105... should not exist because they outdated (  spanning the last 1.5 years ) then re-execute `drop partition` for them.

https://kb.altinity.com/altinity-kb-useful-queries/parts-consistency/#compare-the-list-of-parts-in-zookeeper-with-the-list-of-parts-on-disk
https://kb.altinity.com/upgrade/removing-lost-parts/

21.8 is obsolete, this issue should be resolved automatically in 21.12+ by replacing orphan parts with empty parts.

**Comment by veqryn** [comment]
> > '202104_3043_5805_5_6967','202105_2829_5820_5_7122','202106_0_2841_5_7154'
> 
> Check that this parts really exists in CH (system.parts). Most probably they are orphans which exists in zookeper but not in CH and they should be removed from ZK. If this partitions 202104, 202105... should not exist because they outdated ( spanning the last 1.5 years ) then re-execute `drop partition` for them.
> 
> https://kb.altinity.com/altinity-kb-useful-queries/parts-consistency/#compare-the-list-of-parts-in-zookeeper-with-the-list-of-parts-on-disk https://kb.altinity.com/upgrade/removing-lost-parts/

Yes, those partitions do not exist according to this query:
```sql
select fqdn(), * from clusterAllReplicas(analytics_cluster, system, parts) where `table` = 'xyz_local' and fqdn() = 'clickhouse-us-east-1b-11.something';
```

Why would orphans exist, instead of being completely in-sync with zookeeper?

In 21.12+, does clickhouse automatically remove these orphaned parts from zookeeper?  Or what happens?


> 21.8 is obsolete, this issue should be resolved automatically in 21.12+ by replacing orphan parts with empty parts.

It looks like 21.8.12.29 is an LTS released in December 2021.  Does that mean fixes like cleaning up orphans will be backported to it?

**Comment by den-crane** [comment]
@veqryn emm, did you read https://kb.altinity.com/upgrade/removing-lost-parts/ at all ?

**Comment by den-crane** [comment]
>Why would orphans exist, instead of being completely in-sync with zookeeper?

They are result of the bug.

>In 21.12+, does clickhouse automatically remove these orphaned parts from zookeeper? Or what happens?

21.12+ replaces missing parts with empty parts (0 rows) and then remove empty parts using merge.


>It looks like 21.8.12.29 is an LTS released in December 2021. 
>Does that mean fixes like cleaning up orphans will be backported to it?

It was not backported because it was introduced as a feature not a bugfix.

You need to resolve missing part manually, one time task 

```sql
select 'delete '||part_zoo
from (
select zoo.p_path as part_zoo, zoo.ctime, zoo.mtime, disk.p_path as part_disk
from
(
  select concat(path,'/',name) as p_path, ctime, mtime
  from system.zookeeper where path in (select concat(replica_path,'/parts') from system.replicas)
) zoo
left join 
(
  select concat(replica_path,'/parts/',name) as p_path
  from system.parts inner join system.replicas using (database, table)
) disk on zoo.p_path = disk.p_path
where part_disk='' and zoo.mtime <= now() - interval 1 day
order by part_zoo) format TSVRaw;
```

https://kb.altinity.com/altinity-kb-useful-queries/parts-consistency/#compare-the-list-of-parts-in-zookeeper-with-the-list-of-parts-on-disk

**Comment by den-crane** [comment]
21.8 is out support.

**Comment by veqryn** [comment]
@den-crane So in 21.12+ what happens when updates are run and this happens?

`ALTER TABLE analytics.xyz_local on cluster analytics_cluster UPDATE colA=lower(colA) WHERE colA <> '';`
I just ran the update above, and it froze after finishing everything except for 2 parts on two different nodes.  I checked, and those 2 parts didn't exist anymore, so it was the same or similar to the problem above.

However, I then ran a query and found that the update did not affect ALL the records in the month partition that those 2 parts had belonged to.  **Is this possibility fixed in 21.12+?**

My hypothesis is that those parts got merged with other parts and were skipped over by the update query (ie: you have parts A and B, then you run an update, and the update starts and finishes A, then A and B merge to form C before the update has started B).

**Comment by den-crane** [comment]
It's not fixed in 21.12. It reproduces with 22.3 and 22.8.

Solutions are to clean ZK manually https://kb.altinity.com/altinity-kb-useful-queries/parts-consistency/#compare-the-list-of-parts-in-zookeeper-with-the-list-of-parts-on-disk or to `drop partition` if that partitions are empty or not needed. Not a big deal, it needs only one time.

**Comment by veqryn** [comment]
Correct me if I am wrong, but for the update, the problem isn't a partition that didn't exist, but a partition that didn't get updated. So I am not sure how cleaning zookeeper will help.
Does this bug (updates not affecting all data that they should) already have a ticket/issue, or should I make a new issue for it?

**Comment by den-crane** [comment]
I am talking about this issue: mutation is stuck  `is_done = 0`, `part_to_do = xxxx`, this part xxxx does not exists in the system.parts, but exists in zookeeper, because this record in the zookeeper is a consequence of the bug, it should not exist. All data actually mutated (updated/deleted), mutation can be killed at this point without any consequences because all data has been processed.

**Comment by veqryn** [comment]
> @den-crane So in 21.12+ what happens when updates are run and this happens?
> 
> `ALTER TABLE analytics.xyz_local on cluster analytics_cluster UPDATE colA=lower(colA) WHERE colA <> '';` I just ran the update above, and it froze after finishing everything except for 2 parts on two different nodes. I checked, and those 2 parts didn't exist anymore, so it was the same or similar to the problem above.
> 
> However, I then ran a query and found that the update did not affect ALL the records in the month partition that those 2 parts had belonged to. **Is this possibility fixed in 21.12+?**
> 
> My hypothesis is that those parts got merged with other parts and were skipped over by the update query (ie: you have parts A and B, then you run an update, and the update starts and finishes A, then A and B merge to form C before the update has started B).

In the case above, the mutation did NOT affect ALL the data it should have affected.  Should I make a new ticket/issue for this bug, or should we reopen this one?

**Comment by den-crane** [comment]
Open a new issue.

**Comment by UnamedRus** [comment]
> It's not fixed in 21.12. It reproduces with 22.3 and 22.8 (It happened yesterday with my productionS).

If so, may be we need to reopen issue then?

**Comment by den-crane** [comment]
> > It's not fixed in 21.12. It reproduces with 22.3 and 22.8 (It happened yesterday with my productionS).
> 
> If so, may be we need to reopen issue then?

I meant that is needed to remove orphans from ZK manually once, otherwise mutations will stuck.
